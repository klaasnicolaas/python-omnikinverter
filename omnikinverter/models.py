"""Models for Omnik Inverter."""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any

from .exceptions import OmnikInverterWrongSourceError, OmnikInverterWrongValuesError


@dataclass
class Inverter:
    """Object representing an Inverter response from Omnik Inverter."""

    serial_number: str | None
    model: str | None
    firmware: str | None
    firmware_slave: str | None
    solar_rated_power: int | None
    solar_current_power: int | None
    solar_energy_today: float | None
    solar_energy_total: float | None

    @staticmethod
    def from_json(data: dict[str, Any]) -> Inverter:
        """Return Inverter object from the Omnik Inverter response.

        Args:
            data: The JSON data from the Omnik Inverter.

        Returns:
            An Inverter object.

        Raises:
            OmnikInverterWrongValuesError: Inverter pass on
                incorrect data (day and total are equal).
        """
        data = json.loads(data)

        def get_value(search):
            if data[search] != "":
                return data[search]
            return None

        def validation(data_list):
            """Check if the values are not equal to each other.

            Args:
                data_list: List of values to check.

            Returns:
                Boolean value.
            """
            res = all(ele == data_list[0] for ele in data_list)
            return res

        if validation([data["i_eday"], data["i_eall"]]):
            raise OmnikInverterWrongValuesError(
                "Inverter pass on incorrect data (day and total are equal)"
            )

        return Inverter(
            serial_number=get_value("i_sn"),
            model=get_value("i_modle"),
            firmware=get_value("i_ver_m"),
            firmware_slave=get_value("i_ver_s"),
            solar_rated_power=get_value("i_pow"),
            solar_current_power=int(get_value("i_pow_n")),
            solar_energy_today=float(get_value("i_eday")),
            solar_energy_total=float(get_value("i_eall")),
        )

    @staticmethod
    def from_js(data: dict[str, Any]) -> Inverter:
        """Return Inverter object from the Omnik Inverter response.

        Args:
            data: The JS (webscraping) data from the Omnik Inverter.

        Returns:
            An Inverter object.
        """

        def get_value(position):
            if data.find("webData") != -1:
                matches = re.search(r'(?<=webData=").*?(?=";)', data)
            else:
                matches = re.search(r'(?<=myDeviceArray\[0\]=").*?(?=";)', data)

            try:
                data_list = matches.group(0).split(",")
                if data_list[position] != "":
                    if position in [4, 5, 6, 7]:
                        if position in [4, 5]:
                            return int(data_list[position])

                        if position == 6:
                            energy_value = float(data_list[position]) / 100
                        if position == 7:
                            energy_value = float(data_list[position]) / 10
                        return energy_value
                    return data_list[position].replace(" ", "")
                return None
            except AttributeError as exception:
                raise OmnikInverterWrongSourceError(
                    "Your inverter has no data source from a javascript file."
                ) from exception

        return Inverter(
            serial_number=get_value(0),
            model=get_value(3),
            firmware=get_value(1),
            firmware_slave=get_value(2),
            solar_rated_power=get_value(4),
            solar_current_power=get_value(5),
            solar_energy_today=get_value(6),
            solar_energy_total=get_value(7),
        )


@dataclass
class Device:
    """Object representing an Device response from Omnik Inverter."""

    signal_quality: int | None
    firmware: str | None
    ip_address: str | None

    @staticmethod
    def from_json(data: dict[str, Any]) -> Device:
        """Return Device object from the Omnik Inverter response.

        Args:
            data: The JSON data from the Omnik Inverter.

        Returns:
            An Device object.
        """

        data = json.loads(data)
        return Device(
            signal_quality=None,
            firmware=data["g_ver"].replace("VER:", ""),
            ip_address=data["ip"],
        )

    @staticmethod
    def from_js(data: dict[str, Any]) -> Device:
        """Return Device object from the Omnik Inverter response.

        Args:
            data: The JS (webscraping) data from the Omnik Inverter.

        Returns:
            An Device object.
        """

        def get_value(value_type):
            if value_type == "ip" and data.find("wanIp") != -1:
                match = re.search(r'(?<=wanIp=").*?(?=";)', data.replace(" ", ""))
                value = match.group(0)
            if value_type == "signal" and data.find("m2mRssi") != -1:
                match = re.search(r'(?<=m2mRssi=").*?(?=";)', data.replace(" ", ""))
                value = match.group(0).replace("%", "")
                return int(value)
            if value_type == "version" and data.find("version") != -1:
                match = re.search(r'(?<=version=").*?(?=";)', data.replace(" ", ""))
                value = match.group(0)
            return value

        return Device(
            signal_quality=get_value("signal"),
            firmware=get_value("version"),
            ip_address=get_value("ip"),
        )
