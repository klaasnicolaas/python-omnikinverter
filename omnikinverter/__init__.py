"""Asynchronous Python client for the Omnik Inverter."""

from .models import Device, Inverter, OmnikInverterWrongSourceError
from .omnikinverter import (
    OmnikInverter,
    OmnikInverterConnectionError,
    OmnikInverterError,
)

__all__ = [
    "Device",
    "Inverter",
    "OmnikInverter",
    "OmnikInverterError",
    "OmnikInverterConnectionError",
    "OmnikInverterWrongSourceError",
]
