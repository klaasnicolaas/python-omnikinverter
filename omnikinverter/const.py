"""Constants for the Omnik Inverter."""
import logging

LOGGER = logging.getLogger(__package__)
