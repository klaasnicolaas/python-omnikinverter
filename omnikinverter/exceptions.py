"""Exceptions for Omnik Inverter."""


class OmnikInverterError(Exception):
    """Generic Omnik Inverter exception."""


class OmnikInverterAuthError(OmnikInverterError):
    """Omnik Inverter Authentication exception."""


class OmnikInverterConnectionError(OmnikInverterError):
    """Omnik Inverter connection exception."""


class OmnikInverterWrongSourceError(OmnikInverterError):
    """Omnik Inverter wrong data source url exception."""


class OmnikInverterWrongValuesError(OmnikInverterError):
    """Omnik Inverter gives wrong values.

    This error appears when your inverter
    shows the same value for both day and
    total production.
    """
