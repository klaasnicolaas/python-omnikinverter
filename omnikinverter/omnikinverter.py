"""Asynchronous Python client for the Omnik Inverter."""
from __future__ import annotations

import asyncio
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

import async_timeout
from aiohttp.client import ClientError, ClientResponseError, ClientSession
from yarl import URL

from .exceptions import OmnikInverterConnectionError, OmnikInverterError
from .models import Device, Inverter


@dataclass
class OmnikInverter:
    """Main class for handling connections with the Omnik Inverter."""

    def __init__(
        self,
        host: str,
        use_json: bool = False,
        request_timeout: int = 10,
        session: ClientSession | None = None,
    ) -> None:
        """Initialize connection with the Omnik Inverter.

        Args:
            host: Hostname or IP address of the Omnik Inverter.
            use_json: Boolean to confirm you use a JSON source.
            request_timeout: An integer with the request timeout in seconds.
            session: Optional, shared, aiohttp client session.
        """
        self._session = session
        self._close_session = False

        self.host = host
        self.use_json = use_json
        self.request_timeout = request_timeout

    async def request(
        self,
        uri: str,
        *,
        params: Mapping[str, str] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to a Omnik Inverter device.

        Args:
            uri: Request URI, without '/', for example, 'status'
            params: Extra options to improve or limit the response.

        Returns:
            A Python dictionary (text) with the response from
            the Omnik Inverter.

        Raises:
            OmnikInverterConnectionError: An error occurred while communicating
                with the Omnik Inverter.
            OmnikInverterError: Received an unexpected response from the Omnik Inverter.
        """
        url = URL.build(scheme="http", host=self.host, path="/").join(URL(uri))

        headers = {
            "Accept": "text/html,application/xhtml+xml,application/xml",
        }

        if self._session is None:
            self._session = ClientSession()
            self._close_session = True

        try:
            with async_timeout.timeout(self.request_timeout):
                response = await self._session.request(
                    "GET",
                    url,
                    params=params,
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise OmnikInverterConnectionError(
                "Timeout occurred while connecting to Omnik Inverter device"
            ) from exception
        except (ClientError, ClientResponseError) as exception:
            raise OmnikInverterConnectionError(
                "Error occurred while communicating with Omnik Inverter device"
            ) from exception

        types = ["application/json", "application/x-javascript"]
        content_type = response.headers.get("Content-Type", "")
        if not any(item in content_type for item in types):
            text = await response.text()
            raise OmnikInverterError(
                "Unexpected response from the Omnik Inverter device",
                {"Content-Type": content_type, "response": text},
            )

        return await response.text()

    async def inverter(self) -> Inverter:
        """Get values from your Omnik Inverter.

        Returns:
            A Inverter data object from the Omnik Inverter.
        """
        if self.use_json:
            data = await self.request("status.json", params={"CMD": "inv_query"})
            return Inverter.from_json(data)
        data = await self.request("js/status.js")
        return Inverter.from_js(data)

    async def device(self) -> Device:
        """Get values from the device.

        Returns:
            A Device data object from the Omnik Inverter.
        """
        if self.use_json:
            data = await self.request("status.json", params={"CMD": "inv_query"})
            return Device.from_json(data)
        data = await self.request("js/status.js")
        return Device.from_js(data)

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> OmnikInverter:
        """Async enter.

        Returns:
            The Omnik Inverter object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
