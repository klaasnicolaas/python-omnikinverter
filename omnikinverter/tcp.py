"""Data model and conversions for tcp-based communication with the Omnik Inverter."""
from ctypes import BigEndianStructure, c_char, c_ubyte, c_uint, c_ushort
from typing import Any, Generator, Optional

from .const import LOGGER
from .exceptions import OmnikInverterPacketInvalidError

MESSAGE_START = 0x68
MESSAGE_END = 0x16
MESSAGE_SEND_SEP = 0x40
MESSAGE_RECV_SEP = 0x41
MESSAGE_TYPE_INFORMATION_REQUEST = 0x30
MESSAGE_TYPE_INFORMATION_REPLY = 0xB0
MESSAGE_TYPE_STRING = 0xF0  # Message seems to consist of pure text
UINT16_MAX = 65535
# Message length field does not include the "header":
# - Length (1 byte)
# - Separator (1 byte)
# - Message type (1 byte)
# - Repeated integer serial number (2 * 4 bytes)
# - CRC in suffix (1 byte)
MESSAGE_HEADER_SIZE = 3 + 2 * 4 + 1


class _AcOutput(BigEndianStructure):
    _fields_ = [
        ("frequency", c_ushort),
        ("power", c_ushort),
    ]

    def get_power(self) -> Optional[int]:
        """Retrieve AC power.

        Returns:
            The power field, or None if it is unset.
        """
        return None if self.power == UINT16_MAX else self.power

    def get_frequency(self) -> Optional[float]:
        """Retrieve AC frequency.

        Returns:
            The frequency field in Hertz, or None if it is unset.
        """
        return None if self.frequency == UINT16_MAX else self.frequency * 0.01


class _TcpData(BigEndianStructure):
    _pack_ = 1
    _fields_ = [
        ("padding0", c_char * 3),
        ("serial_number", c_char * 16),
        ("temperature", c_ushort),
        ("dc_input_voltage", c_ushort * 3),
        ("dc_input_current", c_ushort * 3),
        ("ac_output_current", c_ushort * 3),
        ("ac_output_voltage", c_ushort * 3),
        ("ac_output", _AcOutput * 3),
        ("solar_energy_today", c_ushort),
        ("solar_energy_total", c_uint),
        ("solar_hours_total", c_uint),
        ("inverter_active", c_ushort),
        ("padding1", c_ubyte * 4),
        ("unknown0", c_ushort),
        ("padding2", c_ubyte * 10),
        ("firmware", c_char * 16),
        ("padding3", c_ubyte * 4),
        ("firmware_slave", c_char * 16),
        ("padding4", c_ubyte * 4),
    ]


def _pack_message(
    message_type: int, serial_number: int, message: bytearray
) -> bytearray:
    # Prepend message "header"
    request_data = bytearray([len(message), MESSAGE_SEND_SEP, message_type])
    serial_bytes = serial_number.to_bytes(4, "little")
    request_data.extend(serial_bytes)
    request_data.extend(serial_bytes)

    request_data.extend(message)

    checksum = sum(request_data) & 0xFF

    # Finally, prepend message with a start marker, and append the CRC
    # and end marker.
    request_data.insert(0, MESSAGE_START)
    request_data.append(checksum)
    request_data.append(MESSAGE_END)

    return request_data


def _unpack_message(message: bytearray) -> tuple[int, int, bytearray]:
    LOGGER.debug("Handling message `%s`", message)

    message_checksum = message.pop()
    checksum = sum(message) & 0xFF
    if message_checksum != checksum:
        raise OmnikInverterPacketInvalidError(
            f"Checksum mismatch (calculated `{checksum}` got `{message_checksum}`)"
        )

    # Now that the checksum has been computed remove the length,
    # separator, message type and repeated serial number

    length = message.pop(0)  # Length
    if message.pop(0) != MESSAGE_RECV_SEP:
        raise OmnikInverterPacketInvalidError("Invalid receiver separator")

    message_type = message.pop(0)
    LOGGER.debug(
        "Message type %02x, length %s, checksum %02x", message_type, length, checksum
    )

    serial0 = int.from_bytes(message[:4], "little")
    serial1 = int.from_bytes(message[4:8], "little")
    if serial0 != serial1:
        raise OmnikInverterPacketInvalidError(
            f"Serial number mismatch in reply {serial0} != {serial1}"
        )

    return (message_type, serial0, message[8:])


def _unpack_messages(
    data: bytearray,
) -> Generator[tuple[int, int, bytearray], None, None]:
    while len(data):
        message_start = data.pop(0)
        # Whenever my Omnik sends an INFORMATION_REPLY followed by a STRING
        # text message, there's a bunch of trailing 0xFF garbage
        if message_start == 0xFF:  # pragma: no cover
            if not all(d == 0xFF for d in data):
                raise OmnikInverterPacketInvalidError(
                    "(Next) message starts with `0xFF` but the remainder "
                    f"is not strictly 0xFF: {data}"
                )
            # We're done
            return

        if message_start != MESSAGE_START:
            raise OmnikInverterPacketInvalidError("Invalid start byte")

        length = data[0] + MESSAGE_HEADER_SIZE

        message = data[:length]
        if len(message) != length:
            raise OmnikInverterPacketInvalidError(
                f"Could only read {len(message)} out of {length} "
                "expected bytes from TCP stream",
            )

        yield _unpack_message(message)

        # Prepare for the next message by stripping off the end byte
        data = data[length:]
        if data.pop(0) != MESSAGE_END:
            raise OmnikInverterPacketInvalidError("Invalid end byte")


def create_information_request(serial_number: int) -> bytearray:
    """Compute a magic message to which the Omnik will reply with raw statistics.

    Args:
        serial_number: Integer with the serial number of your Omnik device.

    Returns:
        A bytearray with the raw message data, to be sent over a TCP socket.
    """
    return _pack_message(
        MESSAGE_TYPE_INFORMATION_REQUEST, serial_number, bytearray([0x01, 0x00])
    )


def parse_messages(serial_number: int, data: bytes) -> dict[str, Any]:
    """Perform a raw TCP request to the Omnik device.

    Args:
        serial_number: Serial number passed to
            `clk.create_information_request()`, used to validate the reply.
        data: Raw data reply from the Omnik Inverter.

    Returns:
        A Python dictionary (text) with the response from
        the Omnik Inverter.

    Raises:
        OmnikInverterPacketInvalidError: Received data fails basic validity checks.
    """

    info = None

    for (message_type, reply_serial_number, message) in _unpack_messages(
        bytearray(data)
    ):
        if reply_serial_number != serial_number:
            raise OmnikInverterPacketInvalidError(
                f"Replied serial number {reply_serial_number} "
                f"not equal to request {serial_number}"
            )

        if message_type == MESSAGE_TYPE_INFORMATION_REPLY:
            if info is not None:  # pragma: no cover
                LOGGER.warning("Omnik sent multiple INFORMATION_REPLY messages")
            info = _parse_information_reply(message)
        elif message_type == MESSAGE_TYPE_STRING:  # pragma: no cover
            LOGGER.warning(
                "Omnik sent text message `%s`", message.decode("utf8").strip()
            )
        else:
            raise OmnikInverterPacketInvalidError(
                f"Unknown Omnik message type {message_type:02x} "
                f"with contents `{message}`",
            )

    if info is None:
        raise OmnikInverterPacketInvalidError(
            "None of the messages contained an information reply!"
        )

    return info


def _parse_information_reply(data: bytes) -> dict[str, Any]:
    tcp_data = _TcpData.from_buffer_copy(data)

    if tcp_data.unknown0 != UINT16_MAX:  # pragma: no cover
        LOGGER.warning("Unexpected unknown0 `%s`", tcp_data.unknown0)

    if tcp_data.padding0 != b"\x81\x02\x01":  # pragma: no cover
        LOGGER.warning("Unexpected padding0 `%s`", tcp_data.padding0)

    # For all data that's expected to be zero, print it if it's not. Perhaps
    # there are more interesting fields on different inverters waiting to be
    # uncovered.
    for idx in range(1, 5):  # pragma: no cover
        name = f"padding{idx}"
        padding = getattr(tcp_data, name)
        if sum(padding):
            LOGGER.warning("Unexpected `%s`: `%s`", name, padding)

    def list_divide_10(integers: list[int]) -> list[Optional[float]]:
        return [None if v == UINT16_MAX else v * 0.1 for v in integers]

    def int_to_bool(num: int) -> bool:
        return {
            0: False,
            1: True,
        }[num]

    # Only these fields will be extracted from the structure
    field_extractors = {
        "serial_number": None,
        "temperature": 0.1,
        "dc_input_voltage": list_divide_10,
        "dc_input_current": list_divide_10,
        "ac_output_current": list_divide_10,
        "ac_output_voltage": list_divide_10,
        "ac_output": None,
        "solar_energy_today": 0.01,
        "solar_energy_total": 0.1,
        "solar_hours_total": None,
        "inverter_active": int_to_bool,
        "firmware": None,
        "firmware_slave": None,
    }

    result = {}

    for (name, extractor) in field_extractors.items():
        value = getattr(tcp_data, name)
        if name == "ac_output":
            # Flatten the list of frequency+power AC objects

            result["ac_output_frequency"] = [o.get_frequency() for o in value]
            result["ac_output_power"] = [o.get_power() for o in value]
            continue

        if isinstance(extractor, float):
            value *= extractor
        elif extractor is not None:
            value = extractor(value)
        elif isinstance(value, bytes):
            value = value.decode(encoding="utf-8")

        result[name] = value

    return result
