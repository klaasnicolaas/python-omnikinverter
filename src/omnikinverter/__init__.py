"""Asynchronous Python client for the Omnik Inverter."""

from .exceptions import (
    OmnikInverterConnectionError,
    OmnikInverterError,
    OmnikInverterWrongSourceError,
    OmnikInverterWrongValuesError,
)
from .models import Device, Inverter, TcpResponse, WebResponse
from .omnikinverter import OmnikInverter

__all__ = [
    "Device",
    "Inverter",
    "OmnikInverter",
    "OmnikInverterConnectionError",
    "OmnikInverterError",
    "OmnikInverterWrongSourceError",
    "OmnikInverterWrongValuesError",
    "TcpResponse",
    "WebResponse",
]
