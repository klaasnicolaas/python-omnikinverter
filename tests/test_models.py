"""Test the models."""
import aiohttp
import pytest

from omnikinverter import Device, Inverter, OmnikInverter

from . import load_fixtures


@pytest.mark.asyncio
async def test_inverter_js_webdata(aresponses):
    """Test request from a Inverter - JS Webdata source."""
    aresponses.add(
        "example.com",
        "/js/status.js",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/x-javascript"},
            text=load_fixtures("status_webdata.js"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = OmnikInverter(host="example.com", session=session)
        inverter: Inverter = await client.inverter()
        assert inverter
        assert inverter.serial_number == "12345678910"
        assert inverter.firmware == "NL2-V9.8-5931"
        assert inverter.firmware_slave == "V5.3-00157"
        assert inverter.model == "omnik2000tl2"
        assert inverter.solar_current_power == 1010
        assert inverter.solar_energy_today == 4.88
        assert inverter.solar_energy_total == 10531.9


@pytest.mark.asyncio
async def test_device_js_webdata(aresponses):
    """Test request from a Device - JS Webdata source."""
    aresponses.add(
        "example.com",
        "/js/status.js",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/x-javascript"},
            text=load_fixtures("status_webdata.js"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = OmnikInverter(host="example.com", session=session)
        device: Device = await client.device()
        assert device
        assert device.signal_quality == 96
        assert device.firmware == "H4.01.38Y1.0.09W1.0.08"
        assert device.ip_address == "192.168.0.10"


@pytest.mark.asyncio
async def test_inverter_js_devicearray(aresponses):
    """Test request from a Inverter - JS DeviceArray source."""
    aresponses.add(
        "example.com",
        "/js/status.js",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/x-javascript"},
            text=load_fixtures("status_devicearray.js"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = OmnikInverter(host="example.com", session=session)
        inverter: Inverter = await client.inverter()
        assert inverter
        assert inverter.serial_number == "12345678910"
        assert inverter.firmware == "V4.08Build215"
        assert inverter.firmware_slave == "V4.12Build246"
        assert inverter.model == "Omnik1500tl"
        assert inverter.solar_current_power == 850
        assert inverter.solar_energy_today == 2.32
        assert inverter.solar_energy_total == 5200.2


@pytest.mark.asyncio
async def test_device_js_devicearray(aresponses):
    """Test request from a Device - JS DeviceArray source."""
    aresponses.add(
        "example.com",
        "/js/status.js",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/x-javascript"},
            text=load_fixtures("status_devicearray.js"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = OmnikInverter(host="example.com", session=session)
        device: Device = await client.device()
        assert device
        assert device.signal_quality == 39
        assert device.firmware == "H4.01.51MW.2.01W1.0.64(2018-01-251-D)"
        assert device.ip_address == "192.168.0.10"


@pytest.mark.asyncio
async def test_inverter_json(aresponses):
    """Test request from a Inverter - JSON source."""
    aresponses.add(
        "example.com",
        "/status.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("status.json"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        omnik = OmnikInverter(host="example.com", use_json=True, session=session)
        inverter: Inverter = await omnik.inverter()
        assert inverter
        assert inverter.serial_number == "12345678910"
        assert inverter.firmware == "V1.25Build23261"
        assert inverter.firmware_slave == "V1.40Build52927"
        assert inverter.model == "omnik2000tl2"
        assert inverter.solar_current_power == 1225
        assert inverter.solar_energy_today == 10.90
        assert inverter.solar_energy_total == 8674.0


@pytest.mark.asyncio
async def test_device_json(aresponses):
    """Test request from a Inverter - JSON source."""
    aresponses.add(
        "example.com",
        "/status.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("status.json"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        omnik = OmnikInverter(host="example.com", use_json=True, session=session)
        device: Device = await omnik.device()
        assert device
        assert device.signal_quality is None
        assert device.firmware == "ME-111001-V1.0.6(2015-10-16)"
        assert device.ip_address == "192.168.0.10"
