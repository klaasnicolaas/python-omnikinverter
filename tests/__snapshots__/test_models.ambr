# serializer version: 1
# name: test_device_html
  Device(signal_quality=None, firmware='ME_08_0102_2.03', ip_address='192.168.0.106')
# ---
# name: test_device_js_devicearray
  Device(signal_quality=39, firmware='H4.01.51MW.2.01W1.0.64(2018-01-251-D)', ip_address=None)
# ---
# name: test_device_js_webdata
  Device(signal_quality=96, firmware='H4.01.38Y1.0.09W1.0.08', ip_address='192.168.0.10')
# ---
# name: test_device_json
  Device(signal_quality=None, firmware='ME-111001-V1.0.6(2015-10-16)', ip_address='192.168.0.10')
# ---
# name: test_device_without_session
  Device(signal_quality=96, firmware='MW_08_512_0501_1.82', ip_address='192.168.178.3')
# ---
# name: test_inverter_html
  Inverter(serial_number='12345678910', model='Omnik2500tl', solar_rated_power=2500, solar_current_power=219, solar_energy_today=0.23, solar_energy_total=6454.5, alarm_code='F13', firmware='V5.07Build245', firmware_slave=None, inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
# name: test_inverter_js_devicearray
  Inverter(serial_number='12345678910', model='Omnik1500tl', solar_rated_power=None, solar_current_power=850, solar_energy_today=2.32, solar_energy_total=5200.2, alarm_code=None, firmware='V4.08Build215', firmware_slave='V4.12Build246', inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
# name: test_inverter_js_devicearray_sofar2200tl
  Inverter(serial_number='1234567890', model='SOFAR2200TL', solar_rated_power=2000, solar_current_power=400, solar_energy_today=5.67, solar_energy_total=12307.0, alarm_code=None, firmware='V450', firmware_slave=None, inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
# name: test_inverter_js_webdata
  Inverter(serial_number='12345678910', model='omnik4000tl2', solar_rated_power=4000, solar_current_power=140, solar_energy_today=0.3, solar_energy_total=15363.7, alarm_code='F13', firmware='NL2-V9.8-5931', firmware_slave='V5.3-00157', inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
# name: test_inverter_json
  Inverter(serial_number=None, model='omnik2000tl2', solar_rated_power=2000, solar_current_power=1225, solar_energy_today=10.9, solar_energy_total=8674.0, alarm_code='F23', firmware='V1.25Build23261', firmware_slave='V1.40Build52927', inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
# name: test_inverter_without_session
  Inverter(serial_number='1234567890ABCDE', model='0079', solar_rated_power=None, solar_current_power=5850, solar_energy_today=9.8, solar_energy_total=44.0, alarm_code=None, firmware='001F', firmware_slave='002F', inverter_active=None, solar_hours_total=None, temperature=None, dc_input_voltage=None, dc_input_current=None, ac_output_voltage=None, ac_output_current=None, ac_output_frequency=None, ac_output_power=None)
# ---
